### PR TITLE
Remove some low-hanging Guava from core/main

### PR DESCRIPTION
**Pull request by andr3ej**

### CODE REVIEW

**Comment by shamsasari** [comment]
I would suggest you close this PR until Mike's back.

File: core/src/main/kotlin/net/corda/core/Utils.kt [hunk 1]
```diff
@@ -407,3 +406,13 @@ interface DeclaredField<T> {
...
+
+/** The annotated object would have a more restricted visibility were it not needed in tests. */
+@Target(AnnotationTarget.CLASS,
+        AnnotationTarget.PROPERTY,
+        AnnotationTarget.CONSTRUCTOR,
+        AnnotationTarget.FUNCTION,
+        AnnotationTarget.TYPEALIAS)
+@Retention(AnnotationRetention.SOURCE)
+@MustBeDocumented
+annotation class VisibleForTesting
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/Utils.kt]: Bit worried that changing the annotation class makes @VisibleForTesting useless. The problem is people's tools scan for the well known Guava type. Then again this is a very stable part of the Guava framework.

File: core/src/main/kotlin/net/corda/core/utilities/NonEmptySet.kt [hunk 2]
```diff
@@ -50,7 +49,7 @@ class NonEmptySet<T> private constructor(private val elements: Set<T>) : Set<T>
     /** Returns the first element of the set. */
     fun head(): T = elements.iterator().next()
     override fun isEmpty(): Boolean = false
-    override fun iterator(): Iterator<T> = Iterators.unmodifiableIterator(elements.iterator())
+    override fun iterator() = object : Iterator<T> by elements.iterator() {}
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/utilities/NonEmptySet.kt]: what about using Collections.unmodifiableCollection().iterator() ?

File: core/src/main/kotlin/net/corda/core/Utils.kt (hunk 1 above)
Comment (andr3ej) [inline core/src/main/kotlin/net/corda/core/Utils.kt]: Ah OK. Do we use any such tools? If not, I reckon most things visible for testing have not been annotated.

File: core/src/main/kotlin/net/corda/core/utilities/NonEmptySet.kt (hunk 2 above)
Comment (andr3ej) [inline core/src/main/kotlin/net/corda/core/utilities/NonEmptySet.kt]: That would work, but its type is MutableIterator whereas here it's just Iterator.

**Comment by andr3ej** [comment]
The motivation for this PR is that if we have a compile dep of core on guava, guava stuff is likely to sneak back into our API, so how much work would it be to remove that dep? This is the low hanging stuff, there's also some primitives and hashing stuff that would be a bit more work. Will await an executive decision.

**Comment by mikehearn** [comment]
The task is not to remove Guava as a dependency completely, just from the public API. There are two issues with keeping Guava as a dependency:

1. Andrzej's concern that it'll sneak back in by accident. A fair concern, but I'm already planning for us to build a simple tool to do ABI checks to ensure we don't accidentally break our (binary) API in other ways, so throwing in a check for banned dependencies in prototypes isn't that hard.
2. Version conflicts if we get two mutually incompatible versions of Guava into our dependency tree. This is the real killer. But it can be fixed with Jigsaw, at least to some extent, and I'd like us to upgrade to Java 9 some time next year. Until then we can probably struggle through with hacks if this happens.

The upside to keeping it is there's lots of useful stuff in there. It's a good library.

**Pull request merged**